src/java.base/share/classes/jdk/crac/impl/AbstractContextImpl.java

```diff
@@ -22,86 +22,143 @@
 
 package jdk.crac.impl;
 
-import jdk.crac.CheckpointException;
-import jdk.crac.Context;
-import jdk.crac.Resource;
-import jdk.crac.RestoreException;
-import sun.security.action.GetBooleanAction;
+import jdk.crac.*;
+import jdk.internal.crac.LoggerContainer;
 
-import java.util.*;
-import java.util.stream.Collectors;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
 
-public abstract class AbstractContextImpl<R extends Resource, P> extends Context<R> {
+public abstract class AbstractContextImpl<R extends Resource> extends Context<R> {
+    private List<Resource> restoreQ = null;
+    private CheckpointException checkpointException = null;
+    private RestoreException restoreException = null;
 
-    private static class FlagsHolder {
-        public static final boolean DEBUG =
-            GetBooleanAction.privilegedGetProperty("jdk.crac.debug");
+    protected void invokeBeforeCheckpoint(Resource resource) {
+        LoggerContainer.debug("beforeCheckpoint {0}", resource);
+        recordResource(resource);
+        try {
+            resource.beforeCheckpoint(this);
+        } catch (CheckpointException e) {
+            handleCheckpointException(e);
+        } catch (Exception e) {
+            if (e instanceof InterruptedException) {
+                Thread.currentThread().interrupt();
+            }
+            ensureCheckpointException().addSuppressed(e);
+        }
     }
 
-    private WeakHashMap<R, P> checkpointQ = new WeakHashMap<>();
-    private List<R> restoreQ = null;
-    private Comparator<Map.Entry<R, P>> comparator;
+    protected void handleCheckpointException(CheckpointException e) {
+        CheckpointException ce = ensureCheckpointException();
+        for (Throwable t : e.getSuppressed()) {
+            ce.addSuppressed(t);
+        }
+        if (e.getMessage() != null) {
+            ce.addSuppressed(e);
```

Comment:
What happens with exceptions suppressed by `e`? will we have the same set of exceptions suppressed by `ce` and `e`?

Reply:
Yes; there's no way to remove already suppressed exceptions, and the message is too valuable to lose it. I could create an exception with the same message but without suppressed exceptions - then I'd lose its stack trace, though.
I don't see much issues in this duplication.

Reply:
It seems we should have allowed the message in the CheckpointException, to avoid this problem. The only place where CheckpointException with the message is called is recursive checkpoint detection from https://github.com/openjdk/crac/pull/6. 

OK for this PR.

Reply:
Hiding all constructors makes sense. I think it will be helpful to have CheckpointException, CheckpointResourceExceptions and descedants something formal, something that can be programmatically queried for the problem. I propose to investigate the possiblity in #64 which touches related pieces.